lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm

```diff
@@ -119,6 +121,136 @@ sub list {
+
+=back
+
+=cut
+
+sub schedules {
+    my $self = shift;
+
+    my %yaml;
+    my $groups = $self->db->resultset("JobGroups")
+      ->search($self->param('id') ? {id => $self->param('id')} : undef, {select => [qw(id name)]});
+    while (my $group = $groups->next) {
+        my %group;
+        my $templates
+          = $self->db->resultset("JobTemplates")->search({group_id => $group->id}, {order_by => 'me.test_suite_id'});
+
+        # Always set the hash of test suites to account for empty groups
+        $group{architectures} = {};
+        $group{products}      = {};
+
+        my %machines;
+        my %prios;
+        # Extract products and tests per architecture
+        while (my $template = $templates->next) {
+            $group{products}{$template->product->name} = {
+                distri  => $template->product->distri,
+                flavor  => $template->product->flavor,
+                version => $template->product->version
+            };
+            my %test_suite;
+            $test_suite{machine} = $template->machine->name;
+            $machines{$template->product->arch}{$template->machine->name}++;
+            if ($template->prio) {
+                $test_suite{prio} = $template->prio;
+                $prios{$template->product->arch}{$template->prio}++;
+            }
+            my $test_suites = $group{architectures}{$template->product->arch}{$template->product->name};
+            push @$test_suites, {$template->test_suite->name => \%test_suite};
+            $group{architectures}{$template->product->arch}{$template->product->name} = $test_suites;
+        }
+
+        # Split off defaults
+        foreach my $arch (keys %{$group{architectures}}) {
+            # Ensure prios hash is defined for this architecture
+            $prios{$arch}{0} = 0;
+            my %_prios       = %{$prios{$arch}};
+            my $default_prio = (sort { $_prios{$b} <=> $_prios{$a} or $b cmp $a } keys %_prios)[0] + 0;
+            $group{defaults}{$arch}{prio} = $default_prio;
+            my %_machines       = %{$machines{$arch}};
+            my $default_machine = (sort { $_machines{$b} <=> $_machines{$a} or $b cmp $a } keys %_machines)[0];
+            $group{defaults}{$arch}{machine} = $default_machine;
+
+            foreach my $product (keys %{$group{architectures}->{$arch}}) {
+                my @_test_suites;
+                foreach my $test_suite (@{$group{architectures}->{$arch}->{$product}}) {
+                    foreach my $name (keys %$test_suite) {
+                        my %attr = %{$test_suite->{$name}};
+                        if ($attr{machine} eq $default_machine) {
+                            delete $attr{machine};
+                        }
+                        if ($attr{prio} && $attr{prio} == $default_prio) {
+                            delete $attr{prio};
+                        }
+                        if (%attr) {
+                            ${test_suite}->{$name} = \%attr;
```

Comment:
That circumfix operator for dereferencing is not needed here. Use `$test_suite->{$name} = \%attr;`.

Reply:
Maybe try working with references a bit more to get comfortable with them. For example i would have not dereferenced the `%attr` above. To me `my $attr = $test_suite->{$name};` and `$attr->{machine}` look nicer. And you can always use double sigils (`keys %$attr`) if you need to access the hash directly.

Reply:
Slowly getting used to them. Appreciate the advice!